lib/util/itinerary.js

```diff
@@ -245,26 +245,98 @@ export function getLegBounds (leg) {
   return latLngBounds(coords)
 }
 
+/**
+ * Checks whether an appropriate comparison of numeric values can be made for
+ * sorting purposes. If both values are not valid numbers according to the
+ * isNaN check, then this function returns undefined which indicates that a
+ * secondary sorting criteria should be used instead. If one value is valid and
+ * the other is not, then the valid value will be given sorting priority. If
+ * both values are valid numbers, the difference is obtained as the sort value.
+ *
+ * IMPORTANT: both a and b must be numeric values or at least be attempted to
+ * be converted to numeric values! If one of the arguments is something crazy
+ * like an empty string, unexpected behavior will occur because JavaScript.
+ */
+function validNumericValueComparator (a, b) {
+  if (isNaN(a)) {
+    // a is not a number
+    if (isNaN(b)) {
+      // a and b are not numbers, therefore a secondary sorting criteria should
+      // be used. Return undefined to indicate this
+      return undefined
+    }
+    // b is a valid number, b gets priority
+    return 1
+  }
+  // a is a valid number
+  if (isNaN(b)) {
+    // b is not a number, therefore a gets priority
+    return -1
+  }
+  // a and b are valid numbers, return the sort value
+  return a - b
+}
+
+/**
+ * OpenTripPlanner sets the routeSortOrder to -999 by default. So, if that value
+ * is encountered, assume that it actually means that the routeSortOrder is not
+ * set in the GTFS.
+ *
+ * See https://github.com/opentripplanner/OpenTripPlanner/issues/2938
+ * Also see https://github.com/opentripplanner/otp-react-redux/issues/122
+ */
+function getRouteSortOrderValue (val) {
+  return val === -999 ? undefined : val
+}
+
+/**
+ * Compares routes for the purposes of sorting and displaying in a user
+ * interface. Due to GTFS feeds having varying levels of data quality, a
+ * multi-criteria sort is needed to account for various differences. The sort
+ * occurs as follows:
+ * 1. Attempt to sort based off of routeSortOrder. Routes that do not have a
+ *    valid routeSortOrder will be placed beneath those that do.
+ * 2. In case of the same routeSortOrder criteria, attempt to sort based off of
+ *    routeShortNames that are parseable as integers. routeShortNames that
+ *    cannot be parsed as integers will be placed beneath those that are valid.
+ * 3. In case of the same routeShortName criteria, attempt to sort based off of
+ *    the string routeShortName or routeLongName if the routeShortName is
+ *    undefined.
```

Comment:
I'm confused... this JSdoc suggests that it checks the route sort order and then uses route short names to further sort (and so on for short name/long name). Where in the code is that actually handled? It appears that once a valid routeSortOrder comparison is made, the function simply returns.

Reply:
Bah, you're right.

Reply:
Dang, you really had me scratching my head!